### ISSUE TITLE
New validators getting slashed unnecessarily

### ISSUE THREAD

**Issue by mankenavenkatesh** [althea-net/cosmos-gravity-bridge#216]
In valset slashing logic,  

We look through the `currentBondedSet` and slash validators who haven't signed a valset that is currentHeight - signedBlocksWindow old

**Problem:**
if i am a new validator who joined at the end of `SignedValsetsWindow`, i will be slashed and jailed for not confirming the previous validator set which i am not part of.

Also my confirmation doesn't add any value as Peggy contract checks for 2/3+1 confirmations of previous valset request members. 

https://github.com/althea-net/cosmos-gravity-bridge/blob/82dd536d7ce2f7ba5a222eef88a8bd8fab14bd3d/module/x/peggy/abci.go#L40

**Suggestion:**
Instead of checking with `currentBondedSet` we should check with `vs.Members`

**Comment by jkilpatr** [comment]
This is actually a combination of several problems. 

we can't just check vs.Members because of the other side of this issue. When a validator is leaving the set we want them to sign at least a few validator sets that do not contain them (because the set in the bridge *does* contain them). Since we're doing the signing outside of tendermint we have a higher latency, whereas tendermint can change it's entire validator set back to back block after block indefinitely we need to keep validators around and have them sign at least one set that does not contain themselves in order to handle this case.

tl;dr validators who are unbonding need to sign at least one validator set that does not contain them. Using current bondded set is a poor way to represent this but is how it's done currently.

Any ideas as to how we can identify unbonding validators specifically? so that we can force the vs.Members + unbonding validators who unbonded less than x blocks ago to sign

**Comment by mankenavenkatesh** [comment]
One solution that comes to mind is
Listen to `ValsetUpdatedEvent` and broadcast `newValsetNonce` to cosmos. 
Persist Peggy contracts `newValsetNonce` on cosmos.

In endblocker, Check if validators present in Valset mapped with `newValsetNonce` have signed new valset request and slash them.

**Comment by albertchon** [comment]
@mankenavenkatesh's suggestion seems to be the ideal way to do this so that the extra time the unbonding validators leaving the valset have to keep submitting signatures is minimized. However, it introduces yet another flow which may be undesired. 

An alternative (but imperfect) way to pseudo-introduce this flow that would require minimal changes to existing flow could be to emit the `state_lastValsetNonce` inside the `SendToCosmosEvent` which we can then treat as an up-to-date indicator of the current valset status. I think this is okay since in practice, bridge deposits will be a much more frequent occurrence than valset updates, so it seems that in expectation, this should be good enough. 

We'll need to consider the scenarios how this could go wrong in exceptional cases though

**Comment by jkilpatr** [comment]
> @mankenavenkatesh's suggestion seems to be the ideal way to do this so that the extra time the unbonding validators leaving the valset have to keep submitting signatures is minimized. However, it introduces yet another flow which may be undesired.
> 
> An alternative (but imperfect) way to pseudo-introduce this flow that would require minimal changes to existing flow could be to emit the `state_lastValsetNonce` inside the `SendToCosmosEvent` which we can then treat as an up-to-date indicator of the current valset status. I think this is okay since in practice, bridge deposits will be a much more frequent occurrence than valset updates, so it seems that in expectation, this should be good enough.
> 
> We'll need to consider the scenarios how this could go wrong in exceptional cases though

Early on in our design of Gravity we actually had exactly this flow, the latest validator set on Ethereum was stored on Cosmos and we would consider it for computations.

But this is fundamentally unsound from a theory perspective, the current validator set on Ethereum, at any given instant, is unknowable on Cosmos or by anyone. Since ETH is a probabilistic chain there's always a delay before we can declare things final.

On the Cosmos side the validator set can change faster than we can get in signatures for any given update. Imagine for example a scenario where the validator set changes completely back to back every single block indefinitely. This is fully allowed in Tendermint.

Long story short, we don't know what the validator set on Ethereum is, it's unknowable, and at best we can bound the possibilities for the set based on what has signatures and the last known value at some point in history. This bounding isn't as useful as you might imagine though. 

So if we produce validator set B as an update to A the validator set can have already updated to C on the Cosmos chain, leaving Ethereum in a state of "possibly set A or set B". In our imagined scenario of a validator set changing fully every block this goes on to create up to relaying delay / cosmos block time possible states.

We should focus on having enough signed sets that we can guarantee a transition from any one set to any other set through the use of arbitrary intermediate sets is possible.

This is a simple condition, that is instantaneously verifiable at any point in time and is not subject to Ethereum timing issues

The problem with this condition is that we don't produce validator sets every block. If we did we could just have the unbonding validator sign the first block that does not include them and move on.

I believe a 24 hour delay after beginning unbonding where you are still required to sign is totally reasonable. If there's less than 5% cumulative change in that time the validator has to sign no additional sets, if there's greater they sign as many sets as needed.

**Comment by jkilpatr** [comment]
A funny thing to note is that even if you go the 'relay current valset on ethereum to cosmos and use the bounds to constrain possible updates' you still end up with the result of 'a validator has to sign more than one potential update to safely leave the set'.

All roads lead to Rome in that sense.

**Comment by albertchon** [comment]
> We should focus on having enough signed sets that we can guarantee a transition from any one set to any other set through the use of arbitrary intermediate sets is possible
Makes sense, this design is Cosmos-centric in its security approach which in turn lends itself to a more robust security model than what one could seemingly achieve with making assumptions about activity on Ethereum. 

> The problem with this condition is that we don't produce validator sets every block. If we did we could just have the unbonding validator sign the first block that does not include them and move on.
Why not just require the validator to keep signing until the next valset confirmation with 2/3+1 occurs (on the module side)? EDIT oh, that's what you just said with "If there's less than 5% cumulative change in that time the validator has to sign no additional sets". All roads do indeed lead to Rome 🇻🇦

**Comment by mankenavenkatesh** [comment]
Here are my thoughts on identifying which validators to slash and conditions for slashing.
- valsets created before validator bonding. (don't slash validator as his confirmation is not counted)
- valsets created after validator bonded and before validator unbonded. (slash validator)
- valsets created after validator unbonded (slash validator conditionally for a fixed duration)

**Comment by mankenavenkatesh** [comment]
PR for the same - https://github.com/althea-net/cosmos-gravity-bridge/pull/228
